feugy/tabulous
# feat(web): chooses camera and microphone

### What's in there?

It is not unusual for people to have more than one cameras and microphone.
When asking for access, browsers don't let users select which camera/mic to use. It's on the application to enumerate available devices, and display controls for users to choose which ones they'd like to use.

Are included here:

- [feat(web): allows selecting preferred camera/mic](https://github.com/feugy/tabulous/commit/f366b63534ad226b12fc58c5c254e9456742dd0a)
- feat(web): memorizes last camera/mic used
- [feat(web): distinguishes dropdown main and arrow clicks](https://github.com/feugy/tabulous/commit/a57f315e08c15f0d953682e4ff1c893eb8ed0243)
- [feat(web): passes gameId when subscribing to webRTC signals](https://github.com/feugy/tabulous/commit/dc31b3ea3a9d96403b1f9feb9ac247cc93013412)
- [feat(server): updates all players when a peer is joining game](https://github.com/feugy/tabulous/commit/2b2d8b27fc21fc9df1bc04c11596221fb84408b7)
- refactor(web): extract media stream to their own store

### How to test?

Assuming you have several cameras/mics:
1. open a new game
   > the game aside has no players tab
2. invite another player
   > the players tab appears, but stays minimized (new feature)
3. let the other player join
   > the players tab expands to display videos
4. mic & video buttons have drop down arrows
   <img src="https://user-images.githubusercontent.com/186268/172006463-400c01d2-30c7-4278-8865-acbb3d286076.png" width="50%" />
5. now try to switch mic/videos
   > connection is still active, and the other player should see your selected devices
6. refresh the page
   > your last mic/camera should be used.

Please note that, when there's a single choice, the arrow button is not displayed at all.
You can still mute and stop the video independently.
